Introduce StatefulReactView, createStore, and createExternalStateReducer

brettjonesdev opened the pull request:
This PR introduces the new `StatefulReactView`, a lightweight extension of the `ReactView` which works by being passed a Redux `store`, which is renders inside a `<Provider>`. 

## `createStore`
This is a lightweight utility which allows you to quickly and easily create Redux stores. It is a simple convenience method which sets you up with thunk and `extraArgument`. 

## `createExternalStateReducer`
This utility allows you to create a Redux reducer slice which can be controlled externally. The idea is that applications which still have logic that lives inside Backbone models and collections could inject those pieces of state into the store of an application which otherwise relies entirely on Redux to manage its state. See the tests for usage. 

Check out the External State vs Internal State examples in Storybook to learn more! They are identical ToDo Apps, showcasing how the same `StatefulReactView` could be used for a purely Redux application, as well as a "hybrid" application where a piece of the Redux state is externally derived and controlled from outside. Note that their components are essentially identical. The idea would be that an application would be initially written like the External State example, leveraging backbone for some state management, then be refactored to be like the Internal State example, where all state management resides inside Redux. 

## Progression
Inside `hs-app`, we want to slowly replace our Backbone code with React/Redux components. We have many components being controlled by Backbone at a low- or medium-level, and one (Chat) which is a standalone Redux application with a minimal Backbone interface. Over time, we'd like React components which currently exist at a low level inside Backbone applications to move up and consume their Backbone parents, in a kind of [reverse Saturn](https://en.wikipedia.org/wiki/Saturn_Devouring_His_Son). 

After much discussion with @knicklabs, we envision a Backbone -> React transition that looks like this:

### Current
Almost all of our React components living inside hs-app use `ReactMounter` to interface with basic React components. `ReactMounter` is quick and easy to use, but it's very opinionated, and doesn't provide a clear path to transition to being more React-powered. This way of building React components inside Backbone will be considered deprecated moving forward.

We also have a very basic `ReactView` which provides the simplest interface - a single View which wraps a single Component and allows you to pass any props you like into the Component; essentially replacing handlebars with React as a templating engine. 

Lastly, we have the Chat App. The Chat App represents the highest-level of React-ification, short of React-only green field development for new features. Chat App works by having a Marionette Controller initialize a Redux store with preloaded state and `extraArgument`, allowing a tightly controlled point of articulation between React and Backbone. 

### Future
Over time, disparate one-off React components in a Backbone page should give way to larger React portions, and eventually subsume the entire SPA, with `ReactRouter` replacing the Backbone `Router` as the final stage. We picture the progression looking like this: 

* Simple, vanilla React components inside Backbone: `ReactView`
* More complex components leveraging a Redux store, but still backed by state that lives in Backbone models and collections: `StatefulReactView` + `createExternalStateReducer`
* Complex components using Redux, with NO state living in Backbone: `StatefulReactView` 
* Fully React applications: No Backbone, just `ReactRouter`, `Redux` and `React`

brettjonesdev commented:
Note: it's late AF and I haven't updated the documentation yet, but I plan to do so once this PR gets approved and merged up! I'm thinking we'll mark `ReactMounter` deprecated and lay out the new progression.

ItsJonQ commented:
@brettjonesdev Great job on this!!

Watched your video + checked out the code examples.

I think there has to simplify the setup sets for the ExternalState use case:

https://github.com/helpscout/brigade/pull/14/files#diff-9da8b06aefb2414ecb026111b0a6d53eR93

Maybe we can create our own `combineReducers` function, which works like an enhanced version of the one from [Redux](https://redux.js.org/api/combinereducers).

My thinking is something like:

```
const reducers = combineReducers({ ... })
```

Where you can mix regular objects or BB Model/Collections.

The `combineReducers` function will automatically reduce through and detect Backbone things, and rig up the `model.fetch` implementation and things for you.

@knicklabs Thoughts?

brettjonesdev commented:
I like the idea of a `combineReducers`, it would also play nicely with the way we have a custom `createStore` method that wraps around the default Redux `createStore`. I'll start on an implementation, should be pretty easy!

knicklabs commented on docs/05-StatefulReactView.mdx:
```diff
@@ -0,0 +1,167 @@
+
+# StatefulReactView
+
+StatefulReactView is a higher order function (HOF) that ships with Brigade since 1.0.4.
```
Little thing, but lets ship this with `1.1.0`. We should do a minor release for this as its adding a new feature.

brettjonesdev replied:
Great point! Also, I'm not sure how we should update the version number - should we do it in the PR itself, or using an `npm publish minor` or what?

knicklabs replied:
We can update the number here, but you can leave it in package.json. It'll get updated when we publish.

knicklabs commented on src/hoc/ReactMounter/util/buildComponents.js:
```diff
@@ -24,6 +24,7 @@ const buildComponents = (
 ) => {
   return Object.keys(components)
     .map(selector => {
+      // TODO Use StatefulReactView here
```
Can we remove this?

brettjonesdev replied:
done

knicklabs commented on package.json:
```diff
@@ -70,16 +74,24 @@
+    "redux-thunk": "^2.3.0"
   },
   "dependencies": {
+    "redux-mock-store": "^1.5.3",
```
Can we make this a devDependency?

brettjonesdev replied:
doh, yes

brettjonesdev commented on docs/05-StatefulReactView.mdx:
```diff
@@ -0,0 +1,167 @@
+const { reducer, reset } = createExternalReducer('user')
+const reducers = {
+  ...internalReducers,
+  user
```
yes!

brettjonesdev commented on docs/05-StatefulReactView.mdx:
````diff
@@ -0,0 +1,167 @@
+
+### Advanced Usage
+```
+import { createExternalReducer } from '@helpscout/brigade'
````
yes, nice catch

brettjonesdev replied:
Correct! Fixed in latest, pushing shortly.

brettjonesdev commented on src/util/__tests__/createStore.test.js:
```diff
@@ -0,0 +1,81 @@
+  const preloadedState = {count: 5}
+  const reducers = {count}
+  test('should combine reducers and create store capable of incrementing/decrementing count', () => {
+    const store = createStore({preloadedState: {count: 5}, reducers})
```
Yes, done

brettjonesdev commented on src/util/__tests__/createStore.test.js:
```diff
@@ -0,0 +1,81 @@
+    expect(store.getState()).toEqual({count: 4})
+  })
+
+  test('should be able to pass a pre-combined reducer ', () => {
```
fixed! :P

brettjonesdev commented on src/util/__tests__/createStore.test.js:
```diff
@@ -0,0 +1,81 @@
+      doSomething(greeting)
+
+    beforeEach(() => {
+      doSomething = jest.fn()
```
correct, changed

brettjonesdev commented on src/util/createExternalReducer.js:
```diff
@@ -0,0 +1,59 @@
+
+/**
+ * Create a number of external state reducers, automatically binding to collection or model events and dispatching
+ * STATE_RESET_XXX actions to keep them up to date, allowing external control of a portion of Redux state
```
It's actually `RESET_STATE_XXX_TENTACTION`

(too soon?)

brettjonesdev commented:
OK folks, after some conversations with @ItsJonQ I've updated the API just a bit to make things simpler. The `createStore` will now automatically call our `combineReducers` method for you behind the scenes, creating auto-bound external reducers wherever it finds a Model or Collection in the `reducers` map. 

I've updated the docs a bit to make things a bit more clear and highlight the happy path.

TerrenceLJones commented on src/util/__tests__/combineReducers.test.js:
```diff
@@ -0,0 +1,45 @@
+  test('should call createExternalReducers with backbone entitites', () => {
+    const { reducers: reducersReturnVal, unbindExternals: unbindReturnVal } = combineReducers({ todos, user, customers }, getStore)
+    expect(createExternalReducers).toHaveBeenCalledWith({ user, customers }, getStore)
+    expect(createExternalReducers).toHaveBeenCalledWith({ user, customers }, getStore)
```
Duplicate expectation.
